eks/ng/autoscaler/cluster-autoscaler.go

```diff
@@ -0,0 +1,412 @@
+	return &tester{cfg: cfg}
+}
+
+type tester struct {
+	cfg Config
+}
+
+func (ts *tester) Create() error {
+	needInstall := false
+	for _, cur := range ts.cfg.EKSConfig.AddOnNodeGroups.ASGs {
+		if cur.ClusterAutoScaler.Enable {
+			needInstall = true
+			break
+		}
+	}
+	if !needInstall {
+		ts.cfg.Logger.Info("no NG enables CA; skipping")
+		return nil
+	}
+	return ts.createCA()
+}
+
+func (ts *tester) createCA() error {
+	ts.cfg.Logger.Info("creating CA using kubectl", zap.String("name", ts.cfg.EKSConfig.Name))
+	for asgName, cur := range ts.cfg.EKSConfig.AddOnNodeGroups.ASGs {
+		createStart := time.Now()
+		ts.cfg.Logger.Info("creating CA", zap.String("name", asgName))
+		if err := ts.installCA(); err != nil {
+			return err
+		}
+		ts.cfg.Logger.Info("created CA",
+			zap.String("asg-name", cur.Name),
+			zap.String("started", humanize.RelTime(createStart, time.Now(), "ago", "from now")),
+		)
+		ts.cfg.EKSConfig.AddOnNodeGroups.ASGs[asgName] = cur
+		_ = ts.cfg.EKSConfig.Sync()
+	}
+	return ts.cfg.EKSConfig.Sync()
+}
+
+func (ts *tester) installCA() error {
+	var caData = caSpecData{}
+	caData.ImageURI = images[ts.cfg.EKSConfig.Parameters.Version]
+	caData.NodeGroupAutoDiscovery = nodeGroupAuotDiscoveryData + ts.cfg.EKSConfig.Name
+	tpl := template.Must(template.New("TemplateCA").Parse(ClusterAutoscalerYAML))
+	buf := bytes.NewBuffer(nil)
+	if err := tpl.Execute(buf, caData); err != nil {
+		return err
+	}
+	ts.cfg.Logger.Info("writing  cluster autoscaler YAML")
+	fpath, err := fileutil.WriteTempFile(buf.Bytes())
+	if err != nil {
+		ts.cfg.Logger.Warn("failed to write cluster-autoscaler YAML", zap.Error(err))
+		return err
+	}
+	ts.cfg.Logger.Info("applying cluster-autoscaler YAML", zap.String("path", fpath))
+
+	var output []byte
+	waitDur := 5 * time.Minute
+	retryStart := time.Now()
+	applyArgs := []string{
+		ts.cfg.EKSConfig.KubectlPath,
+		"--kubeconfig=" + ts.cfg.EKSConfig.KubeConfigPath,
+		"apply",
+		"-f",
+		fpath,
+	}
+	applyCmd := strings.Join(applyArgs, " ")
+	for time.Now().Sub(retryStart) < waitDur {
+		select {
+		case <-ts.cfg.Stopc:
+			return errors.New("create CA aborted")
+		case <-time.After(5 * time.Second):
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+		output, err := exec.New().CommandContext(ctx, applyArgs[0], applyArgs[1:]...).CombinedOutput()
+		cancel()
+		out := string(output)
+		fmt.Printf("\n\n'%s' output:\n\n%s\n\n", applyCmd, out)
+		if err == nil {
+			break
+		}
+		if strings.Contains(out, " created") || strings.Contains(out, " unchanged") {
+			err = nil
+			break
+		}
+
+		ts.cfg.Logger.Warn("create CA failed", zap.Error(err))
+		ts.cfg.EKSConfig.RecordStatus(fmt.Sprintf("create CA failed (%v)", err))
+	}
+	if err != nil {
+		return fmt.Errorf("'kubectl apply' failed %v (output %q)", err, string(output))
+	}
+	ts.cfg.Logger.Info("created cluster autoscaler")
+
+	return ts.waitDeploymentCA()
+}
+
+func (ts *tester) waitDeploymentCA() error {
```

Comment:
This should take `clusterAutoScalerDeploymentName` as a parameter, in case we run multiple cluster autoscalers for multiple node groups?

Reply:
cluster-autoscaler deployment will only be one per cluster.

Reply:
Ok makes sense.